packages/store/addon/-private/system/store/internal-model-factory.ts

```diff
@@ -0,0 +1,228 @@
+      !(oldId !== null && id !== oldId)
+    );
+
+    // ID can be null if oldID is not null (altered ID in response for a record)
+    // however, this is more than likely a developer error.
+    if (oldId !== null && id === null) {
+      warn(
+        `Your ${modelName} record was saved to the server, but the response does not have an id.`,
+        !(oldId !== null && id === null)
+      );
+      return;
+    }
+
+    let existingInternalModel = this.peekIdOnly(modelName, id);
+
+    assert(
+      `'${modelName}' was saved to the server, but the response returned the new id '${id}', which has already been used with another record.'`,
+      isNone(existingInternalModel) || existingInternalModel === internalModel
+    );
+
+    this.modelMapFor(internalModel.modelName).set(id, internalModel);
+    this._newlyCreatedModelsFor(internalModel.modelName).remove(internalModel, lid);
+
+    internalModel.setId(id);
+  }
+
+  peekIdOnly(modelName: string, id: string): InternalModel | null {
+    let internalModel: InternalModel | null = this.modelMapFor(modelName).get(id);
+
+    if (internalModel && internalModel.hasScheduledDestroy()) {
+      // unloadRecord is async, if one attempts to unload + then sync create,
+      //   we must ensure the unload is complete before starting the create
+      //   The push path will take this.lookup()
+      //   which will call `cancelDestroy` instead for this unload + then
+      //   sync push scenario. Once we have true client-side
+      //   delete signaling, we should never call destroySync
+      internalModel.destroySync();
+      internalModel = null;
+    }
+    return internalModel;
+  }
+
+  build(modelName: string, id: string | null, data?: any, clientId?: string | null) {
+    if (id) {
+      let existingInternalModel = this.peekIdOnly(modelName, id);
+
+      assert(
+        `The id ${id} has already been used with another record for modelClass '${modelName}'.`,
+        !existingInternalModel
+      );
+    }
+
+    if (id === null && !clientId) {
+      clientId = `client-id:${this.newClientId()}`;
+    }
+
+    // lookupFactory should really return an object that creates
+    // instances with the injections applied
+    let internalModel = new InternalModel(modelName, id, this.store, clientId);
+    if (clientId) {
+      this._newlyCreatedModelsFor(modelName).add(internalModel, clientId);
+    }
+
+    this.modelMapFor(modelName).add(internalModel, id);
+
+    return internalModel;
+  }
+
+  newClientId() {
+    return globalClientIdCounter++;
+  }
+
+  remove(internalModel: InternalModel): void {
+    let recordMap = this.modelMapFor(internalModel.modelName);
+    let id = internalModel.id;
+    let clientId = internalModel.clientId;
+
+    if (id) {
+      recordMap.remove(internalModel, id);
+    }
+
+    if (clientId) {
+      this._newlyCreatedModelsFor(internalModel.modelName).remove(internalModel, clientId);
+    }
+  }
+
+  modelMapFor(modelName: string): InternalModelMap {
```

Comment:
can we now refactor this to use this factory/map interface rather than having two things?